File: JavaDemoGithubPractice/app/src/main/java/com/example/javademogithubpractice/ui/base/BaseActivity.java


package com.example.javademogithubpractice.ui.base;

import android.annotation.SuppressLint;
import android.app.ProgressDialog;
import android.content.DialogInterface;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.graphics.Color;
import android.os.Build;
import android.os.Bundle;
import android.os.Handler;
import android.support.annotation.CallSuper;
import android.support.annotation.IdRes;
import android.support.annotation.LayoutRes;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v7.app.AlertDialog;
import android.support.v7.app.AppCompatActivity;
import android.view.Display;
import android.view.MenuItem;
import android.view.View;
import android.view.WindowManager;
import android.widget.Toast;

import com.bumptech.glide.Glide;
import com.example.javademogithubpractice.AppApplication;
import com.example.javademogithubpractice.AppData;
import com.example.javademogithubpractice.dao.DaoSession;
import com.example.javademogithubpractice.inject.component.AppComponent;
import com.example.javademogithubpractice.mvp.contract.IBaseContract;
import com.example.javademogithubpractice.ui.activity.LoginActivity;
import com.example.javademogithubpractice.ui.activity.SplashActivity;
import com.example.javademogithubpractice.util.AppUtils;
import com.example.javademogithubpractice.util.PrefUtils;
import com.example.javademogithubpractice.util.WindowUtil;
import com.thirtydegreesray.dataautoaccess.DataAutoAccess;
import com.thirtydegreesray.dataautoaccess.annotation.AutoAccess;

import java.util.List;

import javax.inject.Inject;

import butterknife.ButterKnife;

public abstract class BaseActivity<P extends IBaseContract.Presenter>
        extends AppCompatActivity implements IBaseContract.View{

    @Inject
    protected P mPresenter;
    private ProgressDialog mProgressDialog;
    private static BaseActivity curActivity;

    protected boolean isAlive = true;
    //@BindView(R.id.toolbar) @Nullable protected Toolbar toolbar;
    //@BindView(R.id.toolbar_layout) @Nullable protected CollapsingToolbarLayout toolbarLayout;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        if((AppData.INSTANCE.getAuthUser() == null || AppData.INSTANCE.getLoggedUser() == null)
                && !this.getClass().equals(SplashActivity.class)
                && !this.getClass().equals(LoginActivity.class)){
            super.onCreate(savedInstanceState);
            finishAffinity();
            startActivity(new Intent(getActivity(), SplashActivity.class));
            return;
        }

        AppUtils.updateAppLanguage(getActivity());
        super.onCreate(savedInstanceState);
        isAlive = true;
        setupActivityComponent(getAppComponent());
        DataAutoAccess.getData(this, savedInstanceState);
        if(mPresenter != null) {
            mPresenter.onRestoreInstanceState(savedInstanceState == null ?
                    getIntent().getExtras() : savedInstanceState);
            mPresenter.attachView(this);
        }
        if(savedInstanceState != null && AppData.INSTANCE.getAuthUser() == null){
            DataAutoAccess.getData(AppData.INSTANCE, savedInstanceState);
        }
        getScreenSize();

        if(getContentView() != 0){
            setContentView(getContentView());
            ButterKnife.bind(getActivity());
        }

        initActivity();
        initView(savedInstanceState);
        if(mPresenter != null) mPresenter.onViewInitialized();
        if(isFullScreen){
            intoFullScreen();
        }
    }

    @Override
    protected void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);
        DataAutoAccess.saveData(this, outState);
        if(mPresenter != null) mPresenter.onSaveInstanceState(outState);
        if(curActivity.equals(this)){
            DataAutoAccess.saveData(AppData.INSTANCE, outState);
        }
    }

    @Override
    protected void onPostCreate(@Nullable Bundle savedInstanceState) {
        super.onPostCreate(savedInstanceState);
    }


    protected abstract void setupActivityComponent(AppComponent appComponent);

    @LayoutRes protected abstract int getContentView();

    @CallSuper
    protected void initActivity(){

    }

    @CallSuper
    protected void initView(Bundle savedInstanceState){
        //if(toolbar != null){
            //setSupportActionBar(toolbar);
//            DoubleClickHandler.setDoubleClickListener(toolbar, new DoubleClickHandler.DoubleClickListener() {
//                @Override
//                public void onDoubleClick(View view) {
//                    onToolbarDoubleClick();
//                }
//            });
        //}
    }

    protected void onToolbarDoubleClick(){
        PrefUtils.set(PrefUtils.DOUBLE_CLICK_TITLE_TIP_ABLE, false);
    }

    @Override
    protected void onResume() {
        super.onResume();
        curActivity = getActivity();
    }

    @Override
    protected void onPause() {
        super.onPause();
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        if(mPresenter != null) mPresenter.detachView();
        if(this.equals(curActivity)) curActivity = null;
        isAlive = false;
    }

    @Override
    public void onTrimMemory(int level) {
        super.onTrimMemory(level);
        Glide.with(this).onTrimMemory(level);
    }

    @Override
    public void onLowMemory() {
        super.onLowMemory();
        Glide.with(this).onLowMemory();
    }

    protected void setToolbarBackEnable() {
        if (getSupportActionBar() != null) {
            getSupportActionBar().setDisplayHomeAsUpEnabled(true);
        }
    }

    protected void setToolbarTitle(String title) {
        if (getSupportActionBar() != null) {
            getSupportActionBar().setTitle(title);
        }
//        if(toolbarLayout != null){
//            toolbarLayout.setTitle(title);
//        }
    }

    protected void setToolbarTitle(String title, String subTitle) {
        setToolbarTitle(title);
        if (getSupportActionBar() != null) {
            getSupportActionBar().setSubtitle(subTitle);
        }
    }

    protected void setToolbarSubTitle(String subTitle) {
        if (getSupportActionBar() != null) {
            getSupportActionBar().setSubtitle(subTitle);
        }
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        if(item.getItemId() == android.R.id.home){
            finishActivity();
            return true;
        }
        return super.onOptionsItemSelected(item);
    }

    public void finishActivity(){
        finish();
    }

    protected Fragment getVisibleFragment(){
        @SuppressLint("RestrictedApi")
        List<Fragment> fragmentList = getSupportFragmentManager().getFragments();
        if(fragmentList != null ){
            for(Fragment fragment : fragmentList){
                if(fragment != null && fragment.isVisible()){
                    return fragment;
                }
            }
        }
        return null;
    }

    @Override
    public void showLoading() {

    }

    @Override
    public void hideLoading() {

    }

    @Override
    public void showProgressDialog(String content) {
        getProgressDialog(content);
        mProgressDialog.show();
    }

    @Override
    public void dismissProgressDialog() {
        if(mProgressDialog != null){
            mProgressDialog.dismiss();
        }else{
            throw new NullPointerException("dismissProgressDialog: can't dismiss a null dialog, must showForRepo dialog first!");
        }
    }

    @Override
    public ProgressDialog getProgressDialog(String content){
        if (mProgressDialog == null) {
            mProgressDialog = new ProgressDialog(getActivity());
            mProgressDialog.setCancelable(false);
        }
        mProgressDialog.setMessage(content);
        return mProgressDialog;
    }

    @Override
    public void showToast(String message){
        Toast.makeText(getActivity(),message,Toast.LENGTH_SHORT).show();
    }

    @Override
    public void showInfoToast(String message) {
        Toast.makeText(getActivity(),message,Toast.LENGTH_SHORT).show();
    }

    @Override
    public void showSuccessToast(String message) {
        Toast.makeText(getActivity(),message,Toast.LENGTH_SHORT).show();
    }

    @Override
    public void showErrorToast(String message) {
        Toast.makeText(getActivity(),message,Toast.LENGTH_SHORT).show();
    }

    @Override
    public void showWarningToast(String message) {
        Toast.makeText(getActivity(),message,Toast.LENGTH_SHORT).show();
    }

    @Override
    public void showTipDialog(String content) {
        new AlertDialog.Builder(getActivity())
                .setCancelable(true)
                .setTitle("qweqwe")
                .setMessage(content)
                .setCancelable(true)
                .setPositiveButton("wqewq", new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(@NonNull DialogInterface dialog, int which) {
                        dialog.cancel();
                    }
                })
                .show();
    }

    @Override
    public void showConfirmDialog(String msn, String title, String confirmText
            , DialogInterface.OnClickListener confirmListener) {
        new AlertDialog.Builder(getActivity())
                .setCancelable(true)
                .setTitle(title)
                .setMessage(msn)
                .setCancelable(true)
                .setPositiveButton(confirmText, confirmListener)
                .setNegativeButton("sadasd", new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(@NonNull DialogInterface dialog, int which) {
                        dialog.cancel();
                    }
                })
                .show();
    }

    protected void postDelayFinish(int time){
        new Handler().postDelayed(new Runnable() {
            @Override
            public void run() {
                finish();
            }
        }, time);
    }

    public static BaseActivity getCurActivity() {
        return curActivity;
    }


    @NonNull
    protected BaseActivity getActivity() {
        return this;
    }

    @NonNull
    protected AppApplication getAppApplication() {
        return (AppApplication) getApplication();
    }

    protected AppComponent getAppComponent(){
        return getAppApplication().getAppComponent();
    }

    protected DaoSession getDaoSession(){
        return getAppComponent().getDaoSession();
    }

    public void onRefreshWebPage(){

    }

    @Nullable
    protected <T extends View> T findViewByViewId(@IdRes int id){
        return (T) findViewById(id);
    }

    private void getScreenSize(){
        if(WindowUtil.screenHeight == 0 ||
                WindowUtil.screenWidth == 0){
            Display display = getWindowManager().getDefaultDisplay();
            WindowUtil.screenWidth = display.getWidth();
            WindowUtil.screenHeight = display.getHeight();
        }
    }

    public String getAppVersionName() {
        String versionName = "";
        try {
            versionName = getPackageManager().getPackageInfo(getPackageName(), 0).versionName;
        } catch (PackageManager.NameNotFoundException e) {
            e.printStackTrace();
        }
        return versionName;
    }

    @NonNull
    protected String getResuceString(int strId) {
        return getResources().getString(strId);
    }

    protected void delayFinish(){
        delayFinish(1000);
    }

    protected void delayFinish(int mills){
        new Handler().postDelayed(new Runnable() {
            @Override
            public void run() {
                finish();
            }
        }, mills);
    }

//    protected void setToolbarScrollAble(boolean scrollAble) {
//        if(toolbar == null) return;
//        int flags = scrollAble ? (AppBarLayout.LayoutParams.SCROLL_FLAG_SCROLL
//                | AppBarLayout.LayoutParams.SCROLL_FLAG_ENTER_ALWAYS
//                | AppBarLayout.LayoutParams.SCROLL_FLAG_SNAP) : 0;
//        AppBarLayout.LayoutParams layoutParams = (AppBarLayout.LayoutParams) toolbar.getLayoutParams();
//        layoutParams.setScrollFlags(flags);
//        toolbar.setLayoutParams(layoutParams);
//    }

    protected void setTransparentStatusBar(){
        if (Build.VERSION.SDK_INT >= 21) {
            View decorView = getWindow().getDecorView();
            int option = View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN | View.SYSTEM_UI_FLAG_LAYOUT_STABLE;
            decorView.setSystemUiVisibility(option);
            getWindow().setStatusBarColor(Color.TRANSPARENT);
        }
    }


    @Override
    public void showLoginPage() {
        getActivity().finishAffinity();
        Intent intent = new Intent(getActivity(), LoginActivity.class);
        startActivity(intent);
    }

    public boolean isAlive() {
        return isAlive;
    }

    @AutoAccess boolean isFullScreen = false;

    protected void exitFullScreen() {
        showStatusBar();
        //if(toolbar != null) toolbar.setVisibility(View.VISIBLE);
        isFullScreen = false;
    }

    protected void intoFullScreen() {
        getWindow().setFlags(WindowManager.LayoutParams.FLAG_FULLSCREEN,
                WindowManager.LayoutParams.FLAG_FULLSCREEN);
        //if(toolbar != null) toolbar.setVisibility(View.GONE);
        isFullScreen = true;
    }

    private void showStatusBar() {
        final WindowManager.LayoutParams attrs = getWindow().getAttributes();
        attrs.flags &= (~WindowManager.LayoutParams.FLAG_FULLSCREEN);
        getWindow().setAttributes(attrs);
        getWindow().clearFlags(WindowManager.LayoutParams.FLAG_LAYOUT_NO_LIMITS);
    }

    @Override
    public void onBackPressed() {
        if(isFullScreen){
            exitFullScreen();
        } else {
            super.onBackPressed();
        }
    }
}

File: KotlinDemoGithubProject/app/src/main/java/com/example/kotlindemogithubproject/ui/base/BaseActivity.kt
package com.example.kotlindemogithubproject.ui.base


import android.app.ProgressDialog
import android.content.DialogInterface
import android.content.Intent
import android.content.pm.PackageManager
import android.os.Bundle
import android.view.MenuItem
import android.view.WindowManager
import android.widget.Toast
import androidx.annotation.CallSuper
import androidx.annotation.LayoutRes
import androidx.annotation.NonNull
import androidx.annotation.Nullable
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import com.bumptech.glide.Glide
import com.example.kotlindemogithubproject.AppApplication
import com.example.kotlindemogithubproject.AppData
import com.example.kotlindemogithubproject.dao.DaoSession
import com.example.kotlindemogithubproject.inject.component.AppComponent
import com.example.kotlindemogithubproject.mvp.contract.IBaseContract
import com.example.kotlindemogithubproject.ui.activity.LoginActivity
import com.example.kotlindemogithubproject.ui.activity.SplashActivity
import com.example.kotlindemogithubproject.util.WindowUtil
import com.thirtydegreesray.dataautoaccess.DataAutoAccess
import com.thirtydegreesray.dataautoaccess.annotation.AutoAccess
import javax.inject.Inject

abstract class BaseActivity<P:IBaseContract.Presenter<*>> : AppCompatActivity(), IBaseContract.View {

    @Inject
    protected var mPresenter: P? = null

    private var mProgressDialog: ProgressDialog? = null

    var isAlive = true


    protected val activity: BaseActivity<*>
        @NonNull
        get() = this

    protected val appApplication: AppApplication
        @NonNull
        get() = application as AppApplication

    protected val appComponent: AppComponent
        get() = appApplication.appComponent!!

    protected val daoSession: DaoSession
        get() = appComponent.daoSession


    val appVersionName: String
        get() {
            var versionName = ""
            try {
                versionName = packageManager.getPackageInfo(packageName, 0).versionName
            } catch (e: PackageManager.NameNotFoundException) {
                e.printStackTrace()
            }

            return versionName
        }

    @AutoAccess
    internal var isFullScreen = false

    override fun onCreate(@Nullable savedInstanceState: Bundle?) {
        if ((AppData.INSTANCE.authUser == null || AppData.INSTANCE.loggedUser == null)
            && this.javaClass != SplashActivity::class.java
            && this.javaClass != LoginActivity::class.java
        ) {
            super.onCreate(savedInstanceState)
            finishAffinity()
            startActivity(Intent(activity, SplashActivity::class.java))
            return
        }

        super.onCreate(savedInstanceState)
        isAlive = true
        setupActivityComponent(appComponent)
        DataAutoAccess.getData(this, savedInstanceState)

        if (mPresenter != null) {
            mPresenter!!.onRestoreInstanceState(savedInstanceState ?: intent.extras)
            mPresenter!!.attachView(this)
        }

        if (savedInstanceState != null && AppData.INSTANCE.authUser == null) {
            DataAutoAccess.getData(AppData.INSTANCE, savedInstanceState)
        }

        getScreenSize()
        if (getContentView() != 0) {
            setContentView(getContentView())
        }

        initActivity()
        initView(savedInstanceState)
        if (mPresenter != null) mPresenter!!.onViewInitialized()
        if (isFullScreen) {
            intoFullScreen()
        }
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        DataAutoAccess.saveData(this, outState)
        if (mPresenter != null) mPresenter!!.onSaveInstanceState(outState)
        if (curActivity == this) {
            DataAutoAccess.saveData(AppData.INSTANCE, outState)
        }
    }

    protected abstract fun setupActivityComponent(appComponent: AppComponent)

    @LayoutRes
    protected abstract fun getContentView(): Int

    @CallSuper
    protected open fun initActivity() {

    }

    @CallSuper
    protected open fun initView(savedInstanceState: Bundle?) {
    }

    override fun onResume() {
        super.onResume()
        curActivity = activity
    }

    override fun onDestroy() {
        super.onDestroy()
        if (mPresenter != null) mPresenter!!.detachView()
        if (this == curActivity) curActivity = null
        isAlive = false
    }

    override fun onTrimMemory(level: Int) {
        super.onTrimMemory(level)
        Glide.with(this).onTrimMemory(level)
    }

    override fun onLowMemory() {
        super.onLowMemory()
        Glide.with(this).onLowMemory()
    }


    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.itemId == android.R.id.home) {
            finishActivity()
            return true
        }
        return super.onOptionsItemSelected(item)
    }

    fun finishActivity() {
        finish()
    }

    override fun showLoading() {

    }

    override fun hideLoading() {

    }

    override fun showProgressDialog(content: String) {
        getProgressDialog(content)
        mProgressDialog!!.show()
    }

    override fun dismissProgressDialog() {
        if (mProgressDialog != null) {
            mProgressDialog!!.dismiss()
        } else {
            throw NullPointerException("dismissProgressDialog: can't dismiss a null dialog, must showForRepo dialog first!")
        }
    }

    override fun getProgressDialog(content: String): ProgressDialog {
        if (mProgressDialog == null) {
            mProgressDialog = ProgressDialog(activity)
            mProgressDialog!!.setCancelable(false)
        }
        mProgressDialog!!.setMessage(content)
        return mProgressDialog as ProgressDialog
    }

    override fun showToast(message: String) {
        Toast.makeText(activity, message, Toast.LENGTH_SHORT).show()
    }

    override fun showInfoToast(message: String) {
        Toast.makeText(activity, message, Toast.LENGTH_SHORT).show()
    }

    override fun showSuccessToast(message: String) {
        Toast.makeText(activity, message, Toast.LENGTH_SHORT).show()
    }

    override fun showErrorToast(message: String) {
        Toast.makeText(activity, message, Toast.LENGTH_SHORT).show()
    }

    override fun showWarningToast(message: String) {
        Toast.makeText(activity, message, Toast.LENGTH_SHORT).show()
    }

    override fun showTipDialog(content: String) {
        AlertDialog.Builder(activity)
            .setCancelable(true)
            .setTitle("qweqwe")
            .setMessage(content)
            .setCancelable(true)
            .setPositiveButton("wqewq", DialogInterface.OnClickListener { dialog, which -> dialog.cancel() })
            .show()
    }

    override fun showConfirmDialog(
        msn: String,
        title: String,
        confirmText: String,
        confirmListener: DialogInterface.OnClickListener
    ) {
        AlertDialog.Builder(activity)
            .setCancelable(true)
            .setTitle(title)
            .setMessage(msn)
            .setCancelable(true)
            .setPositiveButton(confirmText, confirmListener)
            .setNegativeButton("sadasd") { dialog, which -> dialog.cancel() }
            .show()
    }

    private fun getScreenSize() {
        if (WindowUtil.screenHeight === 0 || WindowUtil.screenWidth === 0) {
            val display = windowManager.defaultDisplay
            WindowUtil.screenWidth = display.width
            WindowUtil.screenHeight = display.height
        }
    }

    override fun showLoginPage() {
        activity.finishAffinity()
        val intent = Intent(activity, LoginActivity::class.java)
        startActivity(intent)
    }

    protected fun exitFullScreen() {
        showStatusBar()
        isFullScreen = false
    }

    protected fun intoFullScreen() {
        window.setFlags(
            WindowManager.LayoutParams.FLAG_FULLSCREEN,
            WindowManager.LayoutParams.FLAG_FULLSCREEN
        )
        isFullScreen = true
    }

    private fun showStatusBar() {
        val attrs = window.attributes
        attrs.flags = attrs.flags and WindowManager.LayoutParams.FLAG_FULLSCREEN.inv()
        window.attributes = attrs
        window.clearFlags(WindowManager.LayoutParams.FLAG_LAYOUT_NO_LIMITS)
    }

    override fun onBackPressed() {
        if (isFullScreen) {
            exitFullScreen()
        } else {
            super.onBackPressed()
        }
    }

    companion object {
        var curActivity: BaseActivity<*>? = null
            private set
    }
}
